perf/collector_libpfm.go

```diff
@@ -152,68 +184,145 @@ func (c *collector) setup() error {
-		} else {
-			err = c.setupNonGrouped(string(group[0]), cgroupFd)
+	for i, group := range c.events.Core.Events {
+		// CPUs file descriptors of group leader needed for perf_event_open.
+		leaderFileDescriptors := make(map[int]int, len(c.onlineCPUs))
+		for _, cpu := range c.onlineCPUs {
+			leaderFileDescriptors[cpu] = groupLeaderFileDescriptor
 		}
-		if err != nil {
-			return err
+
+		for j, event := range group.events {
+			// First element is group leader.
+			isGroupLeader := j == 0
+			customEvent, ok := c.eventToCustomEvent[event]
+			if ok {
+				config := c.createConfigFromRawEvent(customEvent, isGroupLeader)
+				leaderFileDescriptors, err = c.registerEvent(eventInfo{string(customEvent.Name), config, cgroupFd, i}, leaderFileDescriptors)
+				if err != nil {
+					return err
+				}
+			} else {
+				config, err := c.createConfigFromEvent(event, isGroupLeader)
+				if err != nil {
+					return err
+				}
+				leaderFileDescriptors, err = c.registerEvent(eventInfo{string(event), config, cgroupFd, i}, leaderFileDescriptors)
+				if err != nil {
+					return err
+				}
+				// Clean memory allocated by C code.
+				C.free(unsafe.Pointer(config))
+			}
+		}
+
+		// Group is prepared so we should reset and enable counting.
+		for _, fd := range leaderFileDescriptors {
+			err = unix.IoctlSetInt(fd, unix.PERF_EVENT_IOC_RESET, 0)
+			if err != nil {
+				return err
+			}
+			err = unix.IoctlSetInt(fd, unix.PERF_EVENT_IOC_ENABLE, 0)
+			if err != nil {
+				return err
+			}
 		}
 	}
 
 	return nil
 }
 
-func (c *collector) setupRawNonGrouped(event *CustomEvent, cgroup int) error {
-	klog.V(5).Infof("Setting up non-grouped raw perf event %#v", event)
-	config := createPerfEventAttr(*event)
-	err := c.registerEvent(config, string(event.Name), cgroup)
-	if err != nil {
-		return err
+func readPerfEventAttr(name string, perfEventAttrMemory unsafe.Pointer) error {
+	event := pfmPerfEncodeArgT{}
+	fstr := C.CString("")
+	event.fstr = unsafe.Pointer(fstr)
+	event.attr = perfEventAttrMemory
+	event.size = C.ulong(unsafe.Sizeof(event))
+	cSafeName := C.CString(name)
+	pErr := C.pfm_get_os_event_encoding(cSafeName, C.PFM_PLM0|C.PFM_PLM3, C.PFM_OS_PERF_EVENT, unsafe.Pointer(&event))
+	if pErr != C.PFM_SUCCESS {
+		return fmt.Errorf("unable to transform event name %s to perf_event_attr: %d", name, int(pErr))
 	}
-
 	return nil
 }
 
-func (c *collector) registerEvent(config *unix.PerfEventAttr, name string, pid int) error {
-	var cpu int
-	for cpu = 0; cpu < c.numCores; cpu++ {
-		groupFd, flags := -1, unix.PERF_FLAG_FD_CLOEXEC|unix.PERF_FLAG_PID_CGROUP
-		fd, err := unix.PerfEventOpen(config, pid, cpu, groupFd, flags)
+type eventInfo struct {
+	name       string
+	config     *unix.PerfEventAttr
+	pid        int
+	groupIndex int
+}
+
+func (c *collector) registerEvent(event eventInfo, leaderFileDescriptors map[int]int) (map[int]int, error) {
+	newLeaderFileDescriptors := make(map[int]int, len(c.onlineCPUs))
+	isGroupLeader := false
```

Comment:
Group leadership if determined in two places:
* here
* https://github.com/google/cadvisor/pull/2578/files#diff-48fe6d1ebf5f72aa1313049c6f1e7aecR196

Can you unify it?

Reply:
Comment about `setAttributes()` is closely related.